src-java/rule-manager/rule-manager-implementation/src/test/java/org/openkilda/rulemanager/factory/generator/flow/haflow/HaFlowRulesBaseTest.java

```diff
@@ -0,0 +1,469 @@
+/* Copyright 2023 Telstra Open Source
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.rulemanager.factory.generator.flow.haflow;
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+import static org.openkilda.model.SwitchFeature.METERS;
+import static org.openkilda.model.SwitchFeature.NOVIFLOW_PUSH_POP_VXLAN;
+import static org.openkilda.model.SwitchFeature.RESET_COUNTS_FLAG;
+import static org.openkilda.rulemanager.Utils.buildSwitch;
+import static org.openkilda.rulemanager.Utils.buildSwitchProperties;
+import static org.openkilda.rulemanager.Utils.buildSwitchPropertiesServer42;
+
+import org.openkilda.model.FlowEncapsulationType;
+import org.openkilda.model.FlowPath;
+import org.openkilda.model.FlowPathDirection;
+import org.openkilda.model.FlowTransitEncapsulation;
+import org.openkilda.model.GroupId;
+import org.openkilda.model.HaFlow;
+import org.openkilda.model.HaFlowPath;
+import org.openkilda.model.HaSubFlow;
+import org.openkilda.model.MeterId;
+import org.openkilda.model.PathId;
+import org.openkilda.model.PathSegment;
+import org.openkilda.model.Switch;
+import org.openkilda.model.SwitchFeature;
+import org.openkilda.model.SwitchId;
+import org.openkilda.model.SwitchProperties;
+import org.openkilda.model.cookie.FlowSegmentCookie;
+import org.openkilda.model.cookie.FlowSubType;
+import org.openkilda.rulemanager.DataAdapter;
+import org.openkilda.rulemanager.FlowSpeakerData;
+import org.openkilda.rulemanager.GroupSpeakerData;
+import org.openkilda.rulemanager.MeterSpeakerData;
+import org.openkilda.rulemanager.OfTable;
+import org.openkilda.rulemanager.SpeakerData;
+import org.openkilda.rulemanager.adapter.InMemoryDataAdapter;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import org.apache.commons.lang3.ArrayUtils;
```

Comment:
Functionality of these libs were added to the standard SDK. Modern, standard function could be optimized better. There is still some difference in style and things like immutable vs unmodifiable, but I'd suggest to use the standard ways where possible.